grpc/optimistic/server.go

```diff
@@ -0,0 +1,308 @@
+					return status.Errorf(codes.Internal, shared.WrapError(err, "error marshalling tx").Error())
+				}
+				marshalledTxs = append(marshalledTxs, marshalledTx)
+
+				bid.Fee = totalCost.Uint64()
+				bid.Transactions = marshalledTxs
+				bid.SequencerParentBlockHash = *o.currentAuctionBlock.Load()
+				bid.RollupParentBlockHash = optimisticBlock.Hash().Bytes()
+
+				txsStreamedCount.Inc(1)
+				err = stream.Send(&auctionPb.GetBidStreamResponse{Bid: &bid})
+				if err != nil {
+					log.Error("error sending bid over stream", "err", err)
+					return status.Error(codes.Internal, shared.WrapError(err, "error sending bid over stream").Error())
+				}
+			}
+
+		case err := <-pendingTxEvent.Err():
+			if err != nil {
+				log.Error("error waiting for pending transactions", "err", err)
+				return status.Error(codes.Internal, shared.WrapError(err, "error waiting for pending transactions").Error())
+			} else {
+				// TODO - what is the right error code here?
+				return status.Error(codes.Internal, "tx pool subscription closed")
+			}
+
+		case <-stream.Context().Done():
+			return stream.Context().Err()
+		}
+	}
+}
+
+func (o *AuctionServiceV1Alpha1) ExecuteOptimisticBlockStream(stream auctionGrpc.OptimisticExecutionService_ExecuteOptimisticBlockStreamServer) error {
+	log.Debug("ExecuteOptimisticBlockStream called")
+
+	mempoolClearingEventCh := make(chan core.NewMempoolCleared)
+	mempoolClearingEvent := o.Eth().TxPool().SubscribeMempoolClearance(mempoolClearingEventCh)
+	defer mempoolClearingEvent.Unsubscribe()
+
+	for {
+		msg, err := stream.Recv()
+		// stream has been closed
+		if errors.Is(err, io.EOF) {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+
+		executeOptimisticBlockRequestCount.Inc(1)
+
+		baseBlock := msg.GetBaseBlock()
+
+		// execute the optimistic block and wait for the mempool clearing event
+		optimisticBlock, err := o.ExecuteOptimisticBlock(stream.Context(), baseBlock)
+		if err != nil {
+			return status.Errorf(codes.Internal, shared.WrapError(err, "failed to execute optimistic block").Error())
+		}
+		optimisticBlockHash := common.BytesToHash(optimisticBlock.Hash)
+
+		// listen to the mempool clearing event and send the response back to the auctioneer when the mempool is cleared
+		select {
+		case event := <-mempoolClearingEventCh:
+			if event.NewHead.Hash() != optimisticBlockHash {
+				log.Error("mempool not cleared after optimistic block execution", "expected_block_hash", optimisticBlockHash.String(), "actual_block_hash", event.NewHead.Hash().String())
+				return status.Error(codes.Internal, "failed to clear mempool after optimistic block execution")
+			}
+			o.currentAuctionBlock.Store(&baseBlock.SequencerBlockHash)
+			executeOptimisticBlockSuccessCount.Inc(1)
+			err = stream.Send(&auctionPb.ExecuteOptimisticBlockStreamResponse{
+				Block:                  optimisticBlock,
+				BaseSequencerBlockHash: baseBlock.SequencerBlockHash,
+			})
+		case <-time.After(500 * time.Millisecond):
```

Comment:
where does this deadline come from?

Reply:
this is kind of arbitrary. 0.5s is too long for the mempool clearance to not take place. We can potentially reduce it too.